lib/livebook/notebook.ex

```diff
@@ -518,12 +536,130 @@ defmodule Livebook.Notebook do
   def find_asset_info(notebook, hash) do
     Enum.find_value(notebook.sections, fn section ->
       Enum.find_value(section.cells, fn cell ->
-        is_struct(cell, Cell.Elixir) &&
-          Enum.find_value(cell.outputs, fn
-            {:js, %{assets: %{hash: ^hash} = assets_info}} -> assets_info
-            _ -> nil
-          end)
+        is_struct(cell, Cell.Elixir) && find_assets_info_in_outputs(cell.outputs, hash)
       end)
     end)
   end
+
+  defp find_assets_info_in_outputs(outputs, hash) do
+    Enum.find_value(outputs, fn
+      {_idx, {:js, %{assets: %{hash: ^hash} = assets_info}}} -> assets_info
+      {_idx, {:frame, outputs, _}} -> find_assets_info_in_outputs(outputs, hash)
+      _ -> nil
+    end)
+  end
+
+  @doc """
+  Adds new output to the given Elixir cell.
+
+  Automatically merges stdout outputs and updates frames.
+  """
+  @spec add_cell_output(t(), Cell.id(), Cell.Elixir.output()) :: t()
+  def add_cell_output(notebook, cell_id, output) do
+    {notebook, counter} = do_add_cell_output(notebook, cell_id, notebook.output_counter, output)
+    %{notebook | output_counter: counter}
+  end
+
+  defp do_add_cell_output(notebook, _cell_id, counter, {:frame, _outputs, %{type: type}} = frame)
+       when type != :default do
+    update_reduce_cells(notebook, counter, fn
+      %Cell.Elixir{} = cell, counter ->
+        {outputs, counter} = update_frames(cell.outputs, counter, frame)
+        {%{cell | outputs: outputs}, counter}
+
+      cell, counter ->
+        {cell, counter}
+    end)
+  end
+
+  defp do_add_cell_output(notebook, cell_id, counter, output) do
+    {output, counter} = index_output(output, counter)
+
+    notebook =
+      update_cell(notebook, cell_id, fn cell ->
+        %{cell | outputs: add_output(cell.outputs, output)}
+      end)
+
+    {notebook, counter}
+  end
+
+  defp update_frames(outputs, counter, {:frame, new_outputs, %{ref: ref, type: type}} = frame) do
+    Enum.map_reduce(outputs, counter, fn
+      {idx, {:frame, outputs, %{ref: ^ref} = info}}, counter ->
+        {new_outputs, counter} = index_outputs(new_outputs, counter)
+        output = {idx, {:frame, apply_frame_update(outputs, new_outputs, type), info}}
+        {output, counter}
+
+      {idx, {:frame, outputs, info}}, counter ->
+        {outputs, counter} = update_frames(outputs, counter, frame)
+        output = {idx, {:frame, outputs, info}}
+        {output, counter}
+
+      output, counter ->
+        {output, counter}
+    end)
+  end
+
+  defp apply_frame_update(_outputs, new_outputs, :replace), do: new_outputs
+  defp apply_frame_update(outputs, new_outputs, :append), do: new_outputs ++ outputs
+
+  defp add_output([], {idx, {:stdout, text}}),
+    do: [{idx, {:stdout, Livebook.Utils.apply_rewind(text)}}]
+
+  defp add_output([], output), do: [output]
+
+  defp add_output(outputs, {_idx, :ignored}), do: outputs
+
+  defp add_output([{_idx1, {:stdout, :__pruned__}} | _] = outputs, {_idx2, {:stdout, _text}}) do
```

Comment:
Isn't the text in this case getting ignored and never added?

Reply:
That's the case where we send update to the component based on the output operation :)

Reply:
So basically if we have binaries we merge, otherwise we keep a single pruned and we know we handle the update separately.

Reply:
What about the clause below then? Doesn't it also send the update and therefore we don't need to apply_rewind?

Reply:
That's what we do in session, where we never prune.

Reply:
Oh, I see. This code is called by both scenarios, so it needs to handle both cases. Can you please add some code comments? I would also move the `prune` function to this module, given that this function is aware of `:__pruned__`, I would say it makes sense to encapsulate this behaviour in this module. :+1:

Reply:
Pruning is quite LV specific, but good point about the encapsulation. Done!